File: packages/frontend/src/types/facilityReservation.ts
import { Timestamp } from "firebase/firestore";
import type { FacilityType } from "./facility";

export type ReservationStatus =
  | "pending" // Awaiting confirmation
  | "confirmed" // Approved and active
  | "rejected" // Admin rejected
  | "cancelled" // User cancelled
  | "completed" // Past reservation
  | "no_show"; // User didn't show up

export interface FacilityReservation {
  id: string;

  // References
  facilityId: string;
  facilityName: string; // Denormalized for display
  facilityType: FacilityType; // Denormalized for filtering
  stableId: string;
  stableName?: string;

  // Reservation details
  userId: string;
  userEmail: string;
  userFullName?: string; // Denormalized for display

  startTime: Timestamp;
  endTime: Timestamp;
  status: ReservationStatus;

  // Horse details - Multi-horse support with backward compatibility
  /** @deprecated Use horseIds instead for new bookings */
  horseId?: string;
  /** @deprecated Use horseNames instead for new bookings */
  horseName?: string; // Denormalized for display

  /** Array of horse IDs for multi-horse bookings (preferred for new bookings) */
  horseIds?: string[];
  /** Array of horse names, denormalized for display (preferred for new bookings) */
  horseNames?: string[];

  /** Number of external horses (not registered in the stable) */
  externalHorseCount?: number;

  // Optional details
  contactInfo?: string;
  notes?: string;

  // Conflict detection
  conflictsWith?: string[]; // Array of overlapping reservation IDs

  // Timestamps (managed by API)
  createdAt: Timestamp;
  updatedAt: Timestamp;
  createdBy: string;
  lastModifiedBy: string;
}

export interface CreateReservationData extends Omit<
  FacilityReservation,
  | "id"
  | "facilityName"
  | "facilityType"
  | "stableId"
  | "stableName"
  | "userEmail"
  | "userFullName"
  | "horseName"
  | "conflictsWith"
  | "createdAt"
  | "updatedAt"
  | "createdBy"
  | "lastModifiedBy"
> {}

export interface UpdateReservationData extends Partial<CreateReservationData> {}

File: EquiDuty-Android/app/src/main/java/com/equiduty/domain/model/FacilityReservation.kt
package com.equiduty.domain.model

data class FacilityReservation(
    val id: String,
    val facilityId: String,
    val facilityName: String,
    val facilityType: FacilityType,
    val stableId: String,
    val stableName: String,
    val userId: String,
    val userEmail: String,
    val userFullName: String,
    val horseId: String?,
    val horseName: String?,
    val horseIds: List<String>?,
    val horseNames: List<String>?,
    val externalHorseCount: Int = 0,
    val startTime: String,
    val endTime: String,
    val purpose: String?,
    val notes: String?,
    val status: ReservationStatus,
    val createdAt: String,
    val updatedAt: String
) {
    /** All horse IDs, normalizing both legacy and new formats */
    val allHorseIds: List<String>
        get() = when {
            !horseIds.isNullOrEmpty() -> horseIds
            !horseId.isNullOrBlank() -> listOf(horseId)
            else -> emptyList()
        }

    /** All horse names, normalizing both legacy and new formats */
    val allHorseNames: List<String>
        get() = when {
            !horseNames.isNullOrEmpty() -> horseNames
            !horseName.isNullOrBlank() -> listOf(horseName)
            else -> emptyList()
        }

    /** Number of horses in this reservation (stable + external) */
    val horseCount: Int get() = allHorseIds.size + externalHorseCount

    /** Display string for horses (single name or count) */
    fun horseDisplayText(pluralFormat: (Int) -> String): String = when (allHorseNames.size) {
        0 -> ""
        1 -> allHorseNames[0]
        else -> pluralFormat(allHorseNames.size)
    }
}

enum class ReservationStatus(val value: String) {
    PENDING("pending"),
    CONFIRMED("confirmed"),
    CANCELLED("cancelled"),
    COMPLETED("completed"),
    NO_SHOW("no_show"),
    REJECTED("rejected");

    companion object {
        fun fromValue(value: String): ReservationStatus =
            entries.find { it.value == value } ?: PENDING
    }
}
